### PR TITLE
Expose ZIP stat information

### PR DESCRIPTION
**Pull request by PVince81**
Required to be able to read the mtime and other metadata.

Note: not added on the interface because it would require adjusting TarStreamer as well

### CODE REVIEW

**Comment by PVince81** [comment]
this is mostly for https://github.com/nextcloud/user_migration/issues/83

but I'm not too happy about the approach there

the elephant in the room is that OC\Archive\Zip::getFiles() is called repeatedly for each call of `getFolder()`, for everything.
maybe I'll optimize that instead and add a method `getMtime($name)` instead

File: lib/private/Archive/ZIP.php [hunk 1]
```diff
@@ -123,6 +123,21 @@ public function getFolder(string $path): array {
...
+	 * @return Generator
+	 */
+	public function getAllFilesStat() {
+		$fileCount = $this->zip->numFiles;
+		$files = [];
+		for ($i = 0;$i < $fileCount;$i++) {
+			yield $this->zip->statIndex($i);
+		}
+
+		return $files;
```
[the remark this answers is not in the document]
Comment (PVince81) [reply lib/private/Archive/ZIP.php:138]: oops, I'll remove it

File: lib/private/Archive/ZIP.php [hunk 2]
```diff
@@ -123,6 +123,28 @@ public function getFolder(string $path): array {
...
 	}
 
+	/**
+	 * Generator that returns metadata of all files
+	 *
+	 * @return Generator
+	 */
+	public function getAllFilesStat() {
+		$fileCount = $this->zip->numFiles;
+		$files = [];
```
Comment (come-nc) [inline lib/private/Archive/ZIP.php:133]:
```suggestion
```

File: lib/private/Archive/ZIP.php [hunk 3]
```diff
@@ -123,6 +123,28 @@ public function getFolder(string $path): array {
 		return $folderContent;
 	}
 
+	/**
+	 * Generator that returns metadata of all files
+	 *
+	 * @return Generator
```
Comment (come-nc) [inline lib/private/Archive/ZIP.php:129]:
```suggestion
	 * @return Generator<int,array>
```

File: lib/private/Archive/ZIP.php [hunk 4]
```diff
@@ -123,6 +126,32 @@ public function getFolder(string $path): array {
...
 	}
 
+	/**
+	 * Generator that returns metadata of all files
+	 *
+	 * @return \Generator<array>
+	 */
+	public function getAllFilesStat() {
+		$fileCount = $this->zip->numFiles;
+		for ($i = 0;$i < $fileCount;$i++) {
```
Comment (Pytal) [inline lib/private/Archive/ZIP.php:136]:
```suggestion
		for ($i = 0; $i < $fileCount; $i++) {
```

**Review by Pytal — APPROVED** [review]

**Comment by Pytal** [comment]
For some reason psalm doesnt find ZipArchive 🤔

File: lib/private/Archive/ZIP.php (hunk 4 above)
Comment (PVince81) [reply lib/private/Archive/ZIP.php:136]: one does not simply copy-paste existing malformatted code

**Comment by come-nc** [comment]
See:

- https://github.com/nextcloud/server/pull/24835
- https://github.com/psalm/psalm-github-actions/issues/27
- https://github.com/psalm/psalm-github-actions/issues/28

TL;DR: it’s not possible / too complicated to have ext-zip installed on the docker for the github psalm action, easier to add a stub for ZipArchive.

**Review by come-nc — APPROVED** [review]
